File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.java
import java.util.*;

class Solution {
    private static final List<Move> moves = new ArrayList<>();
    
    private static class Move {
        public final int start;
        public final int end;
        
        public Move(int start, int end) {
            this.start = start;
            this.end = end;
        }
        
        public int[] toArray() {
            return new int[]{this.start, this.end};
        }
    }
    
    public void move(int n, Move move) {
        final int start = move.start;
        final int end = move.end;
        if (n == 1) {
            moves.add(new Move(start, end));
            return;
        }
        
        // 전반부: 1 ~ n-1번째의 원판을 end가 아닌 곳으로 옮기기
        move(n - 1, new Move(start, 6 - start - end));
        
        // 중반부: n번째 원판을 end로 옮기기
        moves.add(new Move(start, end));
        
        // 후반부: 1 ~ n-1번째 원판을 end로 옮기기
        move(n - 1, new Move(6 - start - end, end));
    }
    
    public int[][] solution(int n) {
        move(n, new Move(1, 3));
        
        int[][] result = new int[moves.size()][2];
        int index = 0;
        for (Move move: moves) {
            result[index++] = move.toArray();
        }
        
        return result;
    }
}

File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.kt
class Solution {
    fun solution(n: Int): Array<IntArray> {
        return move(n, 1, 3)
    }
    
    fun move(m: Int, start: Int, end: Int): Array<IntArray> {
        if (m == 1) {
            return arrayOf(intArrayOf(start, end))
        }
        
        var result = arrayOf<IntArray>()
        move(m - 1, start, mid(start, end)).forEach { result += it }
        result += intArrayOf(start, end)
        move(m - 1, mid(start, end), end).forEach { result += it }
        return result
    }
    
    fun mid(start: Int, end: Int): Int {
        return when (start + end) {
            5 -> 1
            4 -> 2
            3 -> 3
            else -> 0
        }
        // 1 -> 3 = 2
        // 3 -> 1 = 2
        // 1 -> 2 = 3
        // 2 -> 1 = 3
        // 2 -> 3 = 1
        // 3 -> 2 = 1
    }
}

// 원판 1,2,3을 옮기리면
// 1,2를 먼저 옮기고
// 3을 옮기고
// 1을 옮겨야 한다.


// 원판이 1개면 1번 이동
// 원판이 2개면 3번 이동
// 원판이 3개면 7번 이동
// 원판이 4개면 15번?

// 기둥: [A, B, C]
// 원판: [1, 2, 3, 4]
// 규칙을 지키면서 원판을 기둥C로 옮기려면
// 1~3까지를 B로 옮기고
// 4를 C로 옮긴 후
// 1~3를 C로 옮긴다.

// 원판: [1, 2, 3]
// 1~2까지를 B로 옮기고
// 3를 C로 옮긴 후
// 1~3를 C로 옮긴다.

// 원판: [1, 2]
// 1을 B로 옮기고
// 2를 C로 옮긴 후
// 1를 C로 옮긴다.

// 어떤 반복되는 규칙이 있는가
// 시작지점, 임시지점, 목적지점
// 원판이 n개일 때
// 1~n-1번째 원판을 임시지점에 옮기고
// n번쨰 원판을 목적지점에 올리고
// 1~n-1번째 원판을 목적지점에 올리면 됨.
